### ISSUE TITLE
Adding 'pose' information into the gt_boxes

### ISSUE THREAD

**Issue by thanh1985** [jwyang/faster-rcnn.pytorch#592]
Hi,
Thanks for providing the excellent project. 
I am trying not to detect (x,y,w,h) and classify (object name) each bounding box, but also classify object pose (there are 18 poses, each pose corresponds to 10 degrees in a plane). So each object in a image has its label including, object class (hammer, etc), object pose (pose1 (0~10 degrees)), (x,y,w,h).

Take a look at the _ProposalTargetLayer.forward(all_rois, gt_boxes, num_boxes) in proposal_target_layer_cascade.py, the all_rois has shape (1,2000,5), and I notice that all_rois[:,i,:][0][0] is always 0. Because, I already add the pose information into the gt_boxex, now it has shape (1,2000,6(x,y,w,h,class,pose)), so the operation all_rois = torch.cat([all_rois, gt_boxes_append], 1) is not allowed.   

I guess the all_rois (output form RPN) does not have class label information, so I change from gt_boxes_append = gt_boxes.new(gt_boxes.size()).zero_() to gt_boxes_append = gt_boxes.new(torch.Size([1, 20, 5])).zero_() in order to have  all_rois = torch.cat([all_rois, gt_boxes_append], 1) right, though I do not understand the purpose of this.

The problem is that I have to modify the function _sample_rois_pytorch(self, all_rois, gt_boxes, fg_rois_per_image, rois_per_image, num_classes) accordingly, but I do not quite understand.

Could anybody show some hints? Thanks for very much for your time.

**Comment by AlexanderHustinx** [comment]
You could approach it differently and simply pass an additional label with the pose as parameter to the model.
e.g. ```
rois, cls_prob, bbox_pred, \
      rpn_loss_cls, rpn_loss_box, \
      RCNN_loss_cls, RCNN_loss_bbox, \
rois_label = fasterRCNN(im_data, im_info, gt_boxes, num_boxes, gt_poses)
```

And in the `resnet.py` class `resnet(_fasterRCNN)` add an extra one or more modules/layers which you can then call like you do the `RCNN_cls_score` and `RCNN_bbox_pred` in `faster_rcnn.py`, assuming you will base you pose estimation on the `pooled_feat`, similarly to the bbox regression and classification.

**Comment by thanh1985** [comment]
Thanks AlexanderHustinx for the suggestion,
Finally I can get it worked.
However, I find another problem:
For a simple setting:
there are 5 classes (cls1, ..., cls5) and 18 poses (p1, .., p18)
I found that if the Net predicts cls3, then it also predicts p3, (the same index)
My Network is configured the same as Figure 2 (a), in this paper: https://arxiv.org/abs/1801.08110

Do you have any idea about this?

**Comment by AlexanderHustinx** [comment]
Does that mean you've added an extra Linear-module (fc layer) just for the pose? And you connect it just after the `pooled_feat = self._head_to_tail(..)` assignment?

**Comment by thanh1985** [comment]
Hi AlexanderHustinx, yes, it is what I did. I add 
`self.RCNN_ps_score = nn.Linear(4096, self.n_poses)`
and connect it to the `pooled_feat = self._head_to_tail(..)`

**Comment by AlexanderHustinx** [comment]
I believe that should work..

Maybe you forgot to call the `self.RCNN_ps_score(pooled_feat)`?
Which loss function are you using for it?

**Comment by thanh1985** [comment]
I made a mistake, copy the `gt_classes` to  `gt_poses = gt_boxes[:,:,-1]`. Now, it works as I expected. Thanks.
